### ISSUE TITLE
1.0.0-beta.1 fails with Typescript 4.7.4: Attempted to transform a range across two different files

### ISSUE THREAD

**Issue by backspace** [typed-ember/glint#481]
@lukemelia and I tried updating a Yarn [monorepo](https://github.com/cardstack/cardstack) to the `1.0.0-beta.1` today, looking to rid ourselves of spurious warnings like [this](https://github.com/cardstack/cardstack/blob/c8c45e4235c649a3a4ce9b991820f0290386e534/packages/boxel/addon/components/boxel/menu/usage.gts#L6). We encountered a puzzling fatal error running `glint`:

```
Error: Attempted to transform a range across two different files
    at assert (file:///Users/b/Documents/Cardstack/Code/c3/node_modules/@glint/core/lib/transform/util.js:6:15)
    at TransformedModule.getOriginalRange (file:///Users/b/Documents/Cardstack/Code/c3/node_modules/@glint/core/lib/transform/template/transformed-module.js:38:9)
    at rewriteDiagnostic (file:///Users/b/Documents/Cardstack/Code/c3/node_modules/@glint/core/lib/transform/diagnostics/rewrite-diagnostic.js:17:61)
    at TransformManager.rewriteDiagnostic (file:///Users/b/Documents/Cardstack/Code/c3/node_modules/@glint/core/lib/common/transform-manager.js:202:35)
    at TransformManager.rewriteDiagnostics (file:///Users/b/Documents/Cardstack/Code/c3/node_modules/@glint/core/lib/common/transform-manager.js:103:66)
    at performCheck (file:///Users/b/Documents/Cardstack/Code/c3/node_modules/@glint/core/lib/cli/perform-check.js:27:44)
    at file:///Users/b/Documents/Cardstack/Code/c3/node_modules/@glint/core/lib/cli/index.js:108:9
```

That error comes from [this assertion](https://github.com/typed-ember/glint/blob/26fdd02d371d6825de368ecbdd4b5f17ae5bf1f4/packages/core/src/transform/template/transformed-module.ts#L95-L98), which we patched with this helpful logging:

```
        if (startInfo.correlatedSpan.originalFile !== endInfo.correlatedSpan.originalFile) {
          console.log('range error', startInfo.correlatedSpan.originalFile.filename, endInfo.correlatedSpan.originalFile.filename);
        }
```

This revealed that while the files have the same `filename` property, they are not the same objects.

We found that starting with a fresh `yarn.lock` made the problem go away and eventually narrowed it down to the `typescript` package. It was specified as `^4.7.4`, which you can see [failing](https://github.com/cardstack/cardstack/actions/runs/3568953959/jobs/5998390425#step:6:14). Updating to `4.8.0-beta` causes [`glint` to pass](https://github.com/cardstack/cardstack/actions/runs/3568962829/jobs/5998410071#step:6:13).

The [PR](https://github.com/cardstack/cardstack/pull/3465) for the branch that backs those Actions runs is an attempt at a minimal reproduction, if that’s helpful. Let us know if there’s anything else we can supply.

Thanks for all the work on this, now I shudder whenever I return to working with untyped templates 😆

**Comment by dfreeman** [comment]
So! After hunting through this a bit, it turns out https://github.com/typed-ember/glint/commit/ce5a668fd134f84a4bd6e23999cac7f100e7d577 is to blame, and we're just producing a nonsense invalid AST when we try to transform `.gts` files under TS 4.7. Accordingly, TS is producing nonsense invalid diagnostic ranges for those nonsense invalid files, and then we throw up our hands when we can't figure out what to do with some of them.

When making that change I'd wondered if it would have any backwards compatibility issues, but I'd _thought_ our default TS version in the repo was 4.7 and so we were covered. It turns out I mistook a `^` for a `~` and we're actually using 4.8 and doing nothing to ensure 4.7 compatibility 😭 

Realistically I think the path forward is probably that we'll update our minimum supported TS version to 4.8 and document that as a breaking change for 1.0. I'll also fix our version constraint to make sure we don't accidentally drift forward again in the future 😅

**Comment by backspace** [comment]
Makes sense, thanks for looking into this! There was nothing keeping us from updating Typescript other than inertia.
